examples/rag/app_rag_demo.py

```diff
@@ -0,0 +1,479 @@
+        padding-right: 90px;
+        padding-top: 50px;
+        padding-bottom: 50px;
+        background-color: #f0f3f9;
+      }
+    </style>
+  </head>
+  <body>
+    <span class="img-container"> <!-- Inline parent element -->
+      %s
+    </span>
+  </body>
+</html>
+""" % (
+    header_html,
+)
+st.sidebar.markdown(
+    header_full,
+    unsafe_allow_html=True,
+)
+
+# RAG
+description = """
+This demo showcases the [Hugging Face implementation](https://huggingface.co/transformers/master/model_doc/rag.html) of the [RAG model](https://arxiv.org/abs/2005.11401) (for Retrieval Augmented Generation).
+When presented with a query, the model fetches paragraphs from Wikipedia which contain helpful facts and draws information from them to produce an output.
+In addition to providing this output, the present demo helps visualize how these supporting paragraphs are used at each step in the generation.
+"""
+st.sidebar.markdown(description, unsafe_allow_html=True)
+
+action_list = [
+    "Question answering (NaturalQA)",
+    "Question generation (Jeopardy!)",
+]
+
+action_st = st.sidebar.selectbox(
+    "Choose a model setting:",
+    action_list,
+    index=0,
+)
+action_ids = ["nq", "jp"]
+action = action_ids[action_list.index(action_st)]
+
+n_beams = 4
+min_len = 2
+max_len = 64
+
+generate_info = """
+### Answer generation options
+The sequence-to-sequence model uses the [BART](https://huggingface.co/facebook/bart-large)
+architecture. You can increase the beam size to get more answers from the model.
+"""
+generate_options = st.sidebar.checkbox("Generation options")
+if generate_options:
+    st.sidebar.markdown(generate_info)
+    min_len = st.sidebar.slider(
+        "Minimum generation length", min_value=0, max_value=32, value=2, step=2, format=None, key=None
+    )
+    max_len = st.sidebar.slider(
+        "Maximum generation length", min_value=8, max_value=128, value=64, step=8, format=None, key=None
+    )
+    n_beams = st.sidebar.slider("Beam size", min_value=2, max_value=8, value=4, step=2, format=None, key=None)
+
+vis_choice = "weighted"
+# visualization_options = st.sidebar.checkbox("Visualization options")
+visualization_options = False
+if visualization_options:
+    vis_choice = st.sidebar.selectbox(
+        "Choose whether to weigh token-level provenance by the document score:",
+        ["weighted", "un-weighted"],
+        index=0,
+    )
+
+vis_id = 2 if vis_choice == "weighted" else 1
+
+disclaimer = """
+---
+**Disclaimer**
+*The intent of this app is to provide some insights into the behavior of the RAG system.
+Evaluating biases of such a model and ensuring factual generations are still very much open research problems.
+Therefore, until some significant progress is achieved, we caution against using the generated answers for practical purposes.*
+"""
+st.sidebar.markdown(disclaimer, unsafe_allow_html=True)
+
+# Full screen question and answers
+st.markdown("---  ")
+left_a, right_a = st.columns(2)
```

Comment:
The current version of streamlit uses `beta_columns` instead of `columns`. I think this may be from the `horiz-layout` branch of streamlit, so it would be nice to indicate somewhere if that is the case or change to `beta_columns`.

Reply:
It was merged from that branch with the `beta_` prefix in [this PR](https://github.com/streamlit/streamlit/pull/2024)